### ISSUE TITLE
I cannot longer compile with latest version (3.21.0)

### ISSUE THREAD

**Issue by blaart** [elastic/apm-agent-nodejs#2331]
node_modules/elastic-apm-node/index.d.ts:6:33 - error TS2307: Cannot find module 'aws-lambda' or its corresponding type declarations.

6 import type * as AwsLambda from 'aws-lambda';

**Comment by AckerApple** [comment]
Confirming having same issue

**Comment by trentm** [comment]
Thanks. This will be from #2118 (PR: #2119) that (among other things) moved from inlining AwsLambda types to referring to `@types/aws-lambda`. 

Note that that `@types/aws-lambda` was added to *devDependencies* and not *dependencies*. However the same is true of other deps. These are all in devDependencies. 

```
    "@types/aws-lambda": "^8.10.77",
    "@types/connect": "^3.4.34",
    "@types/node": "^13.7.4",
    "@types/pino": "^6.3.8",
```

`@types/node` had been in devDependencies already in previous releases and I gather that was working for folks. I don't know if `@types/node` is a special case or if it is the existing `/// <reference` https://github.com/elastic/apm-agent-nodejs/blob/6547c1e5d75479f20cc83e476f480e9084d10c75/index.d.ts#L1 that makes the difference.

**Comment by trentm** [comment]
@blaart @AckerApple The

```js
import type * as Connect from 'connect';
```

line appears before the AwsLambda import. I'm slightly curious why your error wasn't first from that line. Do you perhaps already have `@types/connect` installed?

```
npm ls @types/connect
```

**Comment by trentm** [comment]
I'd based some of the recent work on fastify's handling of types files. I had not previous seen this: https://github.com/fastify/fastify/blob/e77bfa26809e08e82a5f3180572ccd4d4c4d771a/types/logger.d.ts#L2-L18

It is their current justification for avoiding directly importing `@types/pino` in their types file (and then requiring it to be included in their "dependencies").

I am not leaning to doing the same. I regret having switched to *not* inlining the types from deps that only have their types published as a separate 3rd-party `@types/...`. I will try to get to reverting that soon.

**Comment by AckerApple** [comment]
I have run npm list to give you a dependency tree to study.

I have actually moved to NOT directly importing dependencies because a few times lately its had me stuck and the best solution was to not have direct dependencies. I caution you, that if an additional dependency is needed do NOT make it a direct dependency. Instead put in the README.md install process what's needed such as `npm install aws-lambda  apm-agent-nodejs`

```sh
npm list @types/connect

package-name@1.1.20 /project/path
└─┬ @types/express@4.17.13
  └─┬ @types/body-parser@1.19.1
    └── @types/connect@3.4.35 
```


```
npm list @types/pino

package-name@1.1.20 /Users/ackerapple/projects/bstock/3mp/transactions
└─┬ private-package@0.2.6 (omitted-url-to-private-package)
  └─┬ nestjs-pino@1.4.0
    └─┬ @types/pino-http@5.7.0
      └─┬ @types/pino@6.3.11 
        └─┬ @types/pino-pretty@4.7.1
          └── @types/pino@6.3.11  deduped
```

**Comment by trentm** [comment]
## Repro

```
% cd tsboom

% cat package.json
{
  "name": "tsboom",
  "version": "1.0.0",
  "license": "MIT",
  "dependencies": {
    "elastic-apm-node": "^3.21.0"
  },
  "devDependencies": {
    "typescript": "^4.4.3"
  }
}

% npm install
...

% cat app.ts
import 'elastic-apm-node/start'

% cat tsconfig.json
{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    // "skipLibCheck": true
  }
}

% npx tsc
node_modules/elastic-apm-node/index.d.ts:1:23 - error TS2688: Cannot find type definition file for 'node'.

1 /// <reference types="node" />
                        ~~~~

node_modules/elastic-apm-node/index.d.ts:3:54 - error TS2307: Cannot find module 'http' or its corresponding type declarations.

3 import type { IncomingMessage, ServerResponse } from 'http';
                                                       ~~~~~~

node_modules/elastic-apm-node/index.d.ts:4:43 - error TS7016: Could not find a declaration file for module 'pino'. '/Users/trentm/tmp/tsboom/node_modules/pino/pino.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/pino` if it exists or add a new declaration (.d.ts) file containing `declare module 'pino';`

4 import type { Logger as PinoLogger } from 'pino';
                                            ~~~~~~

node_modules/elastic-apm-node/index.d.ts:5:31 - error TS2307: Cannot find module 'connect' or its corresponding type declarations.

5 import type * as Connect from 'connect';
                                ~~~~~~~~~

node_modules/elastic-apm-node/index.d.ts:6:33 - error TS2307: Cannot find module 'aws-lambda' or its corresponding type declarations.

6 import type * as AwsLambda from 'aws-lambda';
                                  ~~~~~~~~~~~~


Found 5 errors.
```


## Some observations

- A default generated tsconfig.json from `tsc --init` includes `"skipLibCheck": true`, which means a *default* "tsconfig.json" does *not* hit this. I'm guessing that is a default config option because of problems like this being fairly common, but I don't know that.
- A similar repro with Fastify hits similar TS2307 errors because of a missing `@types/node`. The [Fastify docs on TypeScript usage](https://github.com/fastify/fastify/blob/main/docs/TypeScript.md#getting-started) require that the developer separately `npm install --save-dev @types/node`.

## Options

Options to fix this:

1. Put all these `@types/...` devDeps into "dependencies".
2. Remove some or all of the `import type ... from ...` statements and inline types as needed.
3. Require the developer using TypeScript with elastic-apm-node to separately: `npm install --save-dev @types/node @types/...`.

**tl;dr: My initial preference is a form of (2): inline types for all except `@types/node`, assuming this is feasible. [Fastify's justifications for inlining some types](https://github.com/fastify/fastify/blob/e77bfa26809e08e82a5f3180572ccd4d4c4d771a/types/logger.d.ts#L2-L18) apply well here, too.**

* * *

I don't like (1) putting all `@types/...` into "dependencies" because that adds another ~2MiB to the install size of 'elastic-apm-node':

```
% npm install elastic-apm-node
% du -sh node_modules
 14M	node_modules

% npm install @types/aws-lambda @types/connect @types/pino @types/node
% du -sh node_modules
 16M	node_modules

wtsize@1.0.0 /Users/trentm/tmp/wtsize
├── @types/aws-lambda@8.10.83
├─┬ @types/connect@3.4.35
│ └── @types/node@16.9.1 deduped
├── @types/node@16.9.1
├─┬ @types/pino@6.3.11
│ ├── @types/node@16.9.1 deduped
│ ├─┬ @types/pino-pretty@4.7.1
│ │ └── @types/pino@6.3.11 deduped
│ ├─┬ @types/pino-std-serializers@2.4.1
│ │ └── @types/node@16.9.1 deduped
...

% du -sk node_modules/@types/* | sort -n
16	node_modules/@types/connect
16	node_modules/@types/normalize-package-data
20	node_modules/@types/pino-pretty
20	node_modules/@types/pino-std-serializers
132	node_modules/@types/pino
228	node_modules/@types/aws-lambda
1712	node_modules/@types/node
```

Most of that size is from `@types/node`, but aws-lambda is no slouch. The pino and connect types depend on node types.

* * *

I don't love option (3) because most users of 'elastic-apm-node' won't care to know anything about its Pino, Connect, or Lambda facilities. Having to install `@types/node` separate doesn't seem like too much to ask, however.

**Comment by trentm** [comment]
For the record, I briefly considered changing the current:

```typescript
  interface Logger {
    fatal (msg: string, ...args: any[]): void;
    fatal (obj: {}, msg?: string, ...args: any[]): void;
    error (msg: string, ...args: any[]): void;
    error (obj: {}, msg?: string, ...args: any[]): void;
    warn (msg: string, ...args: any[]): void;
    warn (obj: {}, msg?: string, ...args: any[]): void;
    info (msg: string, ...args: any[]): void;
    info (obj: {}, msg?: string, ...args: any[]): void;
    debug (msg: string, ...args: any[]): void;
    debug (obj: {}, msg?: string, ...args: any[]): void;
    trace (msg: string, ...args: any[]): void;
    trace (obj: {}, msg?: string, ...args: any[]): void;
    [propName: string]: any;
  }
```

![Screen Shot 2021-09-16 at 2 47 46 PM](https://user-images.githubusercontent.com/46866/133690917-8a0c676e-483a-4042-84e0-84df1e7cc5a1.png)


to the following which is closer to what fastify and `@types/pino` define:

```typescript
  interface LogFn {
    (msg: string, ...args: unknown[]): void;
    (obj: Record<string, unknown>, msg?: string, ...args: unknown[]): void;
  }
  interface Logger {
    fatal: LogFn;
    error: LogFn;
    warn: LogFn;
    info: LogFn;
    debug: LogFn;
    trace: LogFn;
    [propName: string]: any;
  }
```

![Screen Shot 2021-09-16 at 2 41 51 PM](https://user-images.githubusercontent.com/46866/133690983-9e360d8d-fa54-401f-97d2-8ef9513c8f9f.png)


However, I did not make that change because [as Thomas originally indicated](https://github.com/elastic/apm-agent-nodejs/pull/926/files#r267244795), the completion in VSCode (shown above) is slightly nicer in the former.

**Comment by trentm** [comment]
elastic-apm-node@3.21.1 is published with this fix. Definitely let me know if that doesn't fix it, or if something else comes up. Thanks for the report.
